File: spicedb-java/examples/src/test/java/com/authzed/spicedb/examples/ExpandPermissionTreeTest.java
package com.authzed.spicedb.examples;

import static com.authzed.spicedb.Consistency.*;
import static org.assertj.core.api.Assertions.*;

import com.authzed.spicedb.Filter;
import com.authzed.spicedb.PermissionTree;
import com.authzed.spicedb.PermissionTree.LeafNode;
import com.authzed.spicedb.PermissionTree.Operation;
import com.authzed.spicedb.PermissionTree.SubjectRef;
import com.authzed.spicedb.Relationship;
import com.authzed.spicedb.SpiceDBClient.ExpandResult;
import com.authzed.spicedb.Transaction;
import java.util.HashSet;
import java.util.Set;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

/**
 * Demonstrates expanding a permission into its full tree of subjects using {@link
 * com.authzed.spicedb.SpiceDBClient#expandPermissionTree}, and walking the native {@link
 * PermissionTree} result (expandedObject, expandedRelation, and exactly one of
 * intermediate/leaf) down to its leaf subjects.
 */
class ExpandPermissionTreeTest extends SpiceDBIntegrationTest {

  private String writeRevision;

  @BeforeEach
  void setUp() {
    client.deleteRelationships(Filter.of("document"));

    var txn = new Transaction();
    txn.touch(Relationship.of("document", "report", "viewer", "user", "alice"));
    txn.touch(Relationship.of("document", "report", "editor", "user", "bob"));
    txn.touch(Relationship.of("document", "report", "owner", "user", "charlie"));
    writeRevision = client.write(txn);
  }

  @Test
  void expand_view_permission_returns_union_of_all_relations() {
    ExpandResult result =
        client.expandPermissionTree(atLeast(writeRevision), "document", "report", "view");

    assertThat(result.revision()).isNotEmpty();

    PermissionTree tree = result.tree();
    assertThat(tree.expandedObject().objectType()).isEqualTo("document");
    assertThat(tree.expandedObject().objectID()).isEqualTo("report");
    assertThat(tree.expandedRelation()).isEqualTo("view");

    // "view = viewer + editor + owner" is a union, so the root of the tree is an intermediate
    // node combining the three relations' subtrees.
    assertThat(tree.intermediate()).isNotNull();
    assertThat(tree.leaf()).isNull();
    assertThat(tree.intermediate().operation()).isEqualTo(Operation.UNION);

    Set<String> subjectIDs = collectLeafSubjectIDs(tree);
    assertThat(subjectIDs).containsExactlyInAnyOrder("alice", "bob", "charlie");
  }

  @Test
  void expand_delete_permission_resolves_to_owner_leaf() {
    // "delete = owner" is a single relation with no set operation, which SpiceDB may expand
    // straight to a leaf node rather than wrapping it in a one-child intermediate node. Walking
    // the tree generically (rather than asserting on its exact shape) handles either case.
    ExpandResult result =
        client.expandPermissionTree(atLeast(writeRevision), "document", "report", "delete");

    Set<String> subjectIDs = collectLeafSubjectIDs(result.tree());
    assertThat(subjectIDs).containsExactly("charlie");
  }

  /**
   * Recursively walks a {@link PermissionTree}, collecting subject IDs from every leaf node.
   * Exactly one of {@link PermissionTree#intermediate()} or {@link PermissionTree#leaf()} is
   * non-null on any given node.
   */
  private static Set<String> collectLeafSubjectIDs(PermissionTree tree) {
    Set<String> subjectIDs = new HashSet<>();
    walk(tree, subjectIDs);
    return subjectIDs;
  }

  private static void walk(PermissionTree tree, Set<String> subjectIDs) {
    LeafNode leaf = tree.leaf();
    if (leaf != null) {
      for (SubjectRef subject : leaf.subjects()) {
        subjectIDs.add(subject.subjectID());
      }
      return;
    }

    for (PermissionTree child : tree.intermediate().children()) {
      walk(child, subjectIDs);
    }
  }
}

File: spicedb-csharp/examples/ExpandPermissionTree/ExpandPermissionTreeTest.cs
using Xunit;
// Example ExpandPermissionTree demonstrates expanding a permission into the
// native PermissionTree of subjects with access.

using SpiceDB.Client;
using static SpiceDB.Client.Consistency;

namespace ExpandPermissionTree;

public class ExpandPermissionTreeTest
{
    private const string Schema = """
        definition user {}

        definition document {
            relation viewer: user
            relation editor: user
            permission view = viewer + editor
        }
        """;

    [Fact]
    public async Task ExpandPermissionTree_WalksUnionOfViewerAndEditor()
    {
        await using var client = SpiceDBClient.CreatePlaintext("localhost:50051", "somerandomkeyhere");

        await client.WriteSchemaAsync(Schema);

        var txn = new Transaction();
        txn.Touch(Relationship.FromTriple("document", "report", "viewer", "user", "alice"));
        txn.Touch(Relationship.FromTriple("document", "report", "editor", "user", "bob"));
        var revision = await client.WriteAsync(txn);

        var result = await client.ExpandPermissionTreeAsync(
            AtLeast(revision), "document", "report", "view");

        // The root node identifies the resource/permission that was expanded.
        Assert.Equal("document", result.Tree.ExpandedObject.ObjectType);
        Assert.Equal("report", result.Tree.ExpandedObject.ObjectID);
        Assert.Equal("view", result.Tree.ExpandedRelation);

        // "view = viewer + editor" is a union, so the root is an
        // IntermediateNode with a Union operation and no Leaf.
        Assert.Null(result.Tree.Leaf);
        Assert.NotNull(result.Tree.Intermediate);
        Assert.Equal(TreeOperation.Union, result.Tree.Intermediate!.Operation);

        // Walk the tree recursively, collecting every subject found at a leaf.
        var subjectIDs = CollectLeafSubjects(result.Tree)
            .Select(s => s.SubjectID)
            .ToHashSet();

        Assert.Contains("alice", subjectIDs);
        Assert.Contains("bob", subjectIDs);
    }

    /// <summary>
    /// Recursively walks a <see cref="PermissionTree"/>, yielding every
    /// subject found at a leaf node. Exactly one of Intermediate/Leaf is
    /// non-null on each node, so both cases must be handled.
    /// </summary>
    private static IEnumerable<SubjectRef> CollectLeafSubjects(PermissionTree tree)
    {
        if (tree.Leaf is { } leaf)
        {
            foreach (var subject in leaf.Subjects)
            {
                yield return subject;
            }
        }

        if (tree.Intermediate is { } intermediate)
        {
            foreach (var child in intermediate.Children)
            {
                foreach (var subject in CollectLeafSubjects(child))
                {
                    yield return subject;
                }
            }
        }
    }
}
